### DISCUSSION TITLE
The way to run async functions in Celey tasks in one event loop

### DISCUSSION THREAD

**Discussion by YuriiMotov** [celery/celery#9058] [Q&A]
Hi everyone!

I have some async code and I need to run it inside the Celery task.

I tried the approach with using `asgiref.sync.async_to_sync()`, but it turned out that it creates new event loop every time. And it brakes my code since I use SQLAlchemy session pool and there are restrictions about using sessions in different threads and event loops.

**It is important to note** that I don't care about performance issues. I understand the overheads of using this approach. I just don't want to rewrite my async code to sync.

Trying to run all my async functions in one event loop I created a simple helper module:

```
import asyncio
import time
from collections.abc import Awaitable
from threading import Thread
from typing import TypeVar


class _SingleEventLoop:
    _loop: asyncio.AbstractEventLoop | None = None
    _loop_thread: Thread | None = None

    def _enshure_loop_is_running(self):
        if self._loop and (self._loop.is_running() is False):
            self._loop.close()
            self._loop = None
            if self._loop_thread.is_alive():
                self._loop_thread.stop()

        if (self._loop is None) or (self._loop.is_running() is False):
            self._loop_thread = Thread(target=self._eventloop_thread_run, daemon=True)
            self._loop_thread.start()
            time.sleep(0.1)

    def _eventloop_thread_run(self):
        self._loop = asyncio.new_event_loop()
        asyncio.set_event_loop(self._loop)
        self._loop.run_forever()

    def execute_async_task(self, coroutine):
        self._enshure_loop_is_running()
        feature = asyncio.run_coroutine_threadsafe(coroutine, self._loop)
        return feature.result()


_single_event_loop = _SingleEventLoop()

R = TypeVar("R")

def execute_async_task(coroutine: Awaitable[R]) -> R:
    """Executes async tasks in single event loop"""
    return _single_event_loop.execute_async_task(coroutine=coroutine)
```

I can execute my async code like it's shown below:

```
from single_event_loop_runner.run_async import execute_async_task

@celery_app.task()
def sync_celery_task(param):
    res = execute_async_task(async_func(param))
    return res

```

It seems to work fine for my use case, but I want to know what are the possible negative consequences of using this approach?
Can I use any easier way to solve my task?

I'm quite new to Celery and need your help!)

**UPD:** updated code to use separate event loop (not to use event loop that might be already running in the current thread).

**Comment by Roman-Ka** [comment]
@YuriiMotov Thanks a lot for posting this, this solved a problem I was banging my head against the wall on for 3 days

**Reply by Roman-Ka** [reply]
@YuriiMotov one question I have is your line `self._loop_thread.stop()`, I only found `_stop()` method on the `Thread` object. I am using python 3.12. Am I missing something?

**Reply by YuriiMotov** [reply]
Yes, this is a mistake..
I think you can just remove these two lines:
```
            if self._loop_thread.is_alive():
                self._loop_thread.stop()
```

**Comment by OleksandrZhydyk** [comment]
Found another option to resolve this issue, the engine object needs to be disposed after each task call.

https://stackoverflow.com/a/73253766

**Comment by albertedwardson** [comment]
Hi @YuriiMotov, @Roman-Ka —
Since this issue seems to come up often, I’ve built a small library that cleanly bridges async from sync using a persistent background event loop: [palitra](https://github.com/abebus/palitra)

It’s lightweight, no monkey-patching, and should help with cases like this. PRs and feedback are very welcome
